src/commons/application/ApplicationTypes.ts

```diff
@@ -157,7 +165,7 @@ export const htmlLanguage: SALanguage = {
   mainLanguage: SupportedLanguage.JAVASCRIPT
 };
 
-const schemeSubLanguages = [
+const schemeSubLanguages: Array<Pick<SALanguage, 'chapter' | 'variant' | 'displayName'>> = [
```

Comment:
This makes the typing unnecessarily convoluted. :/

Could `displayName` be part of `Language` instead? It doesn't make sense that the Source sublanguages are handled differently from Scheme and Python in this regard.

Alternatively, the display names could be derived for Scheme and Python in the same manner as for Source. The way Scheme and Python currently do it is absolutely terrible because the `styliseSublanguage` function is the source of truth for the Source derived names, but not for Scheme and Python:
https://github.com/source-academy/frontend/blob/ce58e6d5d3a2db9719957065156a70ec4a43a604/src/commons/application/ApplicationTypes.ts#L184-L208
Instead, we get the above hacks that make use of the non-null assertion operator. (We really should ban it in the linter so that people stop adding them...)

To be honest, I'm not sure why when it comes to displaying these language strings, we sometimes call `styliseSublanguage` and other times just access the `displayName` attribute. If they are truly one and the same, we should make use of the latter only, with the former possibly remaining as a convenience function for calculating all of the Source sublanguage display names.

Reply:
> Could `displayName` be part of `Language` instead?

I agree this should be the case. Unfortunately `Language` is defined in js-slang, not the frontend. Perhaps we can create our own version of `Language` temporarily? That way we can use `SALanguage` $<:$ `Language` (frontend) $<:$ `Language` (js-slang).

Regarding `styliseSublanguage`, I wholeheartedly agree. I wasn't sure if I should do the refactoring for that in this PR as well. Essentially, I believe that what is happening is that originally, when we only had Source, they tried to abstract it this way such that one can get the display name simply by passing a chapter and a variant, which might have seemed like a good idea then, but now it's just messy since we have languages that are not Source, so in an effort to add support for these languages for now, they did this patchwork.

Here's my proposal:

1. Move https://github.com/source-academy/frontend/blob/93396dc689e743e5f517e813a93e7ffe520c8b6d/src/commons/application/ApplicationTypes.ts#L212-L213 in `styliseSublanguage` to https://github.com/source-academy/frontend/blob/93396dc689e743e5f517e813a93e7ffe520c8b6d/src/commons/application/ApplicationTypes.ts#L263
2. (Temporarily) rewrite `styliseSublanguage` to simply return `getLanguageConfig(chapter, variant).displayName`: https://github.com/source-academy/frontend/blob/93396dc689e743e5f517e813a93e7ffe520c8b6d/src/commons/application/ApplicationTypes.ts#L280-L292 for compatibility
3. Once logic fully migrated away from simply passing around the chapter and variant (see last task of #2463), we replace every instance of `styliseSublanguage` to the `displayName` attribute directly

What do you think?

Reply:
> I agree this should be the case. Unfortunately `Language` is defined in js-slang, not the frontend. Perhaps we can create our own version of `Language` temporarily? That way we can use `SALanguage` $<:$ `Language` (frontend) $<:$ `Language` (js-slang).

That makes sense.

> Regarding `styliseSublanguage`, I wholeheartedly agree. I wasn't sure if I should do the refactoring for that in this PR as well. Essentially, I believe that what is happening is that originally, when we only had Source, they tried to abstract it this way such that one can get the display name simply by passing a chapter and a variant, which might have seemed like a good idea then, but now it's just messy since we have languages that are not Source, so in an effort to add support for these languages for now, they did this patchwork.

Yea, I agree that this is slightly out of scope of this PR already. Unfortunately, it also unnecessarily complicates the typing here.

> Here's my proposal:
>
> 1. Move https://github.com/source-academy/frontend/blob/93396dc689e743e5f517e813a93e7ffe520c8b6d/src/commons/application/ApplicationTypes.ts#L212-L213 in `styliseSublanguage` to https://github.com/source-academy/frontend/blob/93396dc689e743e5f517e813a93e7ffe520c8b6d/src/commons/application/ApplicationTypes.ts#L263
> 2. (Temporarily) rewrite `styliseSublanguage` to simply return `getLanguageConfig(chapter, variant).displayName`: https://github.com/source-academy/frontend/blob/93396dc689e743e5f517e813a93e7ffe520c8b6d/src/commons/application/ApplicationTypes.ts#L280-L292 for compatibility
> 3. Once logic fully migrated away from simply passing around the chapter and variant (see last task of #2463), we replace every instance of `styliseSublanguage` to the `displayName` attribute directly
>
> What do you think?

Sounds great! Also sounds very painful... Thanks for carrying out this refactoring.

Reply:
Steps 1 and 2 have been done in 7850dbce629649614093631ca90ccd5027a21746 as part of #2470.